Sharding and other updates

infinisil opened the pull request:
- Describe sharding as discussed in the last NAT meeting
- Refines which packages are eligible for this transition
  - Exclude functions and packages that aren't self-contained regarding references
- Add some TODO's
- General text improvements
- Add an example tree

Closes #16, closes #19

roberth commented on README.md:
```diff
@@ -11,51 +11,61 @@ related-issues: (will contain links to implementation PRs)
 
-These attributes will newly be defined by listing all directories in `pkgs/auto` using `builtins.readDir` and calling `pkgs.callPackage` on all of them, which then gets added to the `pkgs` scope.
+## Compatibility layer
+[compatibility]: #compatibility-layer
```
This may be more trouble than it's worth.

infinisil replied:
I don't think it's much trouble, it doesn't take a lot of effort to implement

roberth replied:
Any trouble at all is too much trouble of the effect isn't positive. By doing this we set a precedent that any file can be a public interface, while they're not supposed to.

SuperSandro2000 replied:
The file layout was not a public interface before. We just move packages and paths around with no considerations of breakages.

blaggacao replied:
I sort of agree with the sentiment of cost/benefit questions.

I think the main benefit is for people with am established muscle memory to quickly be redirected for the first couple of runs unil the muscle memory has adapted.

A one-line text file without any functionality other than informing the new location could just do.

No strong opinion, but maybe a way to scope this down a little.

infinisil replied:
We discussed this [in yesterdays NAT meeting](https://discourse.nixos.org/t/2022-11-28-nixpkgs-architecture-team-meeting-19/23572#simple-package-paths-4). We concluded that we don't want a compatibility layer at all, because:
- The filesystem generally is not an API
  - While there's `pkgs.path`, it doesn't really make sense with overlays and config
  - There's `modulesPath` but that's specific to NixOS modules
  - There's `meta.position`, but only for debugging/better errors
- Especially with [the requirement for files to not be changed, only moved](https://github.com/nixpkgs-architecture/simple-package-paths/pull/20/files#r1035249558), not having a compatibility symlink works really nicely with PR's, as we believe git can handle rename merges properly

infinisil replied:
Now removed the compatibility layer section, I added a mention to it in the alternatives section instead

roberth commented on README.md:
```diff
@@ -104,38 +147,30 @@ This RFC makes no requirement as to how the transition should happen, but here a
-   and we want to improve tooling to make renames easy.
-
+- Use a flat directory `pkgs/unit/*/pkg-fun.nix` instead, arguments:
+  - Good because it speeds up Nix evaluation since there's only a single directory to call `builtins.readDir` on instead of many
+    - With an optimized `readDir` this isn't much of a problem
+  - Good because it's simpler, both for the user and for the code
```
Swap.

```suggestion
  - Good because it's simpler, both for the user and for the code
  - Good because it speeds up Nix evaluation since there's only a single directory to call `builtins.readDir` on instead of many
    - With an optimized `readDir` this isn't much of a problem
```

roberth commented on README.md:
```diff
@@ -71,13 +81,48 @@ This RFC makes no requirement as to how the transition should happen, but here a
+   │  ├── autosuspend
+   │  ┊
+   ┊
+   ├── sl--
```
Why not `substring 0 4 "sl"`?

Likewise for zsh.

aakropotkin replied:
Allow me to paint a picture. 

Howard has been using Nix for one year. He excitedly makes his first PR to Nixpkgs.

The path names seem weird :

"foo/foo/default.nix"
"u/u/default.nix"
"emac/emacs/default.nix"

Howard is confused about the max length but tries his best, he PRs "libxz/libxz/default.nix". Howard gets an email requesting that he refilehis PR with a new name, he rolls his eyes "how was I supposed to guess that?", but tells himself he'll refile when he has some free time.

It is 6 months later. The seasons have changed, but Howard's PR has not. Stale Bot adds yet another notification to Robert's feed of stale PRs - left dying on the vine.

Nixpkgs is in shambles! If only the maintainers had chosen `jq--/jq/default.nix` this all could have been avoided!

</scene>

infinisil replied:
[Another comment by @aakropotkin on Discourse](https://discourse.nixos.org/t/2022-11-21-nixpkgs-architecture-team-meeting-18/23393/4?u=infinisil)

Originally in the meeting we used `--` in case there's multiple levels of sharding, like

```
he/ll/hello
zs/h-/zsh
sl/--/sl
r-/--/r
```

This makes it clear that e.g. `sl` shouldn't be in `sl/sl/pkg-fun.nix` where it would collide with a package named `slsl` in `sl/sl/slsl/pkg-fun.nix`. `sl` being in `sl/--/sl/pkg-fun.nix` gets rid of the ambiguity.

However, later in the meeting we figured out that there's no good reason to have nested folders, so we changed it to 
```
```
hell/hello
zsh-/zsh
sl--/sl
r---/r
```

instead, which lines up very nicely still. But you're right roberth, with this scheme there's no super good reason for the `-` anymore.

I'm 50/50 on this, because as @aakropotkin mentions, the `-` do help with clarity, establishing that it's a 4-letter prefix, and not just the package name. Of course we could enforce either option with CI checks (and we totally should).

roberth replied:
> The path names seem weird :
> 
> "foo/foo/default.nix"
> "u/u/default.nix"
> "emac/emacs/default.nix"

This is not realistic.
Only 1 in 22 packages is shorter than 4 characters, and only 1 in 8 isn't truncated.
Let's avoid solving problems that don't exist.

> CI checks (and we totally should).

I agree. We don't want to have any drift from the sharding specification, because that would defeat the point.

SuperSandro2000 replied:
IMO filling up shorter packages with dashes is ugly. Can't we just trim them to a minimal length and keep them as is if they are not longer?

infinisil replied:
@SuperSandro2000 Yes that's @roberth's proposal we're discussing here

infinisil replied:
I updated the PR to use this proposal now

SuperSandro2000 commented on README.md:
```diff
@@ -104,38 +147,30 @@ This RFC makes no requirement as to how the transition should happen, but here a
+    - With an optimized `readDir` this isn't much of a problem
+  - Good because it's simpler, both for the user and for the code
+  - Bad because it causes GitHub to limit the rendering of that directory to 1'000 entries (and we have about 10'000 that benefit from this transition for a start)
+  - Bad because it makes `git` slower ([TODO: By how much?](https://github.com/nixpkgs-architecture/simple-package-paths/issues/18))
```
```suggestion
  - Bad because it makes `git` slower ([TODO: By how much?](https://github.com/nixpkgs-architecture/simple-package-paths/issues/18))
  - Bad because directory listing slows down with many files
```

This is the main reason why most object stores and CDNs prefix their hash paths with directories ab/cd/efg.... .

SuperSandro2000 commented on README.md:
```diff
@@ -11,51 +11,61 @@ related-issues: (will contain links to implementation PRs)
+When moving `pkgs/some/dir/default.nix` to the new `pkgs/unit/<4-prefix name>/<name>/pkg-fun.nix`, a symlink will be created pointing from the old to the new location. Reasoning:
 - Current community discussions referencing old files from the `master` branch are still valid for some time. While GitHub doesn't provide an easy way to navigate to a symlink, seeing the path to where the file has moved is better than getting an error.
 - It provides an opportunity for code referencing old paths to be updated. While it's not possible to give a deprecation warning with symlinks, users will at least be able to read it in the NixOS release notes. This doesn't occur often in practice.
 
```
Code that does not expect a symlink due to bad writing will break. Probably acceptable.

How to make sure we do not collect dangling symlinks even in that one NixOS release? If anything gets moved afterwards I don't expect people to think about the symlink myself including.

infinisil replied:
That's a good point, this increases the complexity of this compatibility layer by a lot. What we could do is limit it to exactly one release and remove it right after. This way we only need to check/correct for dead symlinks/Nix files once when the branch-off happens and can remove them in master afterwards. But yes this adds some complexity.

infinisil replied:
Discussed in https://github.com/nixpkgs-architecture/simple-package-paths/pull/20#discussion_r1028999091

SuperSandro2000 commented on README.md:
```diff
@@ -11,51 +11,61 @@ related-issues: (will contain links to implementation PRs)
-{
-  syncplay-nogui = syncplay.override { enableGUI = false; };
-}
+builtins.trace "warning: Using deprecated path ${./.}, use pkgs/unit/<name> instead, this will be removed after NixOS 22.05"
+  (import ../../pkgs/unit/name)
```
That does not work with ofborg and potentially spams people evaluation log.

infinisil replied:
No it shouldn't, since nothing in nixpkgs evaluates these files anymore at that point.

SuperSandro2000 commented on README.md:
```diff
@@ -11,51 +11,61 @@ related-issues: (will contain links to implementation PRs)
-  - Which directory should my package definition go in? What are all the categories and do they matter? What if the package has multiple matching categories?
-  - Why can't I build my package after adding the package file? [introduced to all-packages.nix] Where in all-packages.nix should my package go?
-- Figuring out where an attribute is defined is very tricky:
+- (Especially new) package contributors are having a hard time figuring out which files to add or edit. These are very common questions:
```
I am not sure if that will improve the situation for new people. They will still have questions and they will just be different.

infinisil replied:
It will certainly improve the situation! It removes some of the most annoying problems to deal with when contributing changes to packages. And yes this won't fix all the problems, just one of the most prominent ones. And just because there's always more problems doesn't mean solving one won't improve the situation.

SuperSandro2000 commented on README.md:
```diff
@@ -11,51 +11,61 @@ related-issues: (will contain links to implementation PRs)
-    - Especially on GitHub this is even more problematic, as the `all-packages.nix` file is [too big to be displayed by GitHub](https://github.com/NixOS/nixpkgs/blob/nixos-22.05/pkgs/top-level/all-packages.nix)
+    - On GitHub this is even more problematic, as the `all-packages.nix` file is [too big to be displayed by GitHub](https://github.com/NixOS/nixpkgs/blob/nixos-22.05/pkgs/top-level/all-packages.nix)
   - Then go to that file's definition, which takes quite some time for navigation (unless you have a plugin that can jump to it directly)
 - `all-packages.nix` frequently causes merge conflicts. It's a point of contention for all new packages
```
We should mention nix edit somewhere. This is the main way I get around the packages and is pretty easy and works for most packages very well.

infinisil replied:
I'm mentioning `nix edit` in this section now

SuperSandro2000 commented on README.md:
```diff
@@ -11,51 +11,61 @@ related-issues: (will contain links to implementation PRs)
+- Figuring out where an attribute is defined is a bit tricky:
   - First one has to find the definition of it in all-packages.nix to see what file it refers to
-    - Especially on GitHub this is even more problematic, as the `all-packages.nix` file is [too big to be displayed by GitHub](https://github.com/NixOS/nixpkgs/blob/nixos-22.05/pkgs/top-level/all-packages.nix)
+    - On GitHub this is even more problematic, as the `all-packages.nix` file is [too big to be displayed by GitHub](https://github.com/NixOS/nixpkgs/blob/nixos-22.05/pkgs/top-level/all-packages.nix)
```
also local editors slow down due to file size and a badly written nix language server can deadlock when editing.

infinisil replied:
Added a mention of local editors being slow

infinisil commented on README.md:
```diff
@@ -11,51 +11,61 @@ related-issues: (will contain links to implementation PRs)
+For all attributes at the root of nixpkgs `pkgs.<name>` which:
+1. Are defined in `pkgs/top-level/all-packages.nix`
+  (necessary so that the overlay containing the automatically discovered packages can be ordered directly before the `all-packages.nix` overlay without changing any behavior)
+2. Are defined to be equal to `pkgs.callPackage <path> { }` such that all transitively referenced paths from the default Nix file of `<path>`:
+  1. Are under the directory of `<path>`
+    (necessary so that moving these files to a new directory doesn't break references in this package)
+  2. Are not referenced from any paths outside of these transitive references
+    (necessary so that moving these files to a new directory doesn't break references in other packages)
```
We discussed this [in yesterdays NAT meeting](https://discourse.nixos.org/t/2022-11-28-nixpkgs-architecture-team-meeting-19/23572#relative-file-references-8). We concluded that this is a good idea, but it's hard to understand what's written here, it should be reworded to something like "Limited to packages that can be moved without updating any file references"

infinisil replied:
The tooling could also be regularly ran in CI, giving warnings for all packages that can't be moved automatically, along with an indication for why it couldn't be moved automatically. This is especially a list of special cases we need to take a look at then

infinisil replied:
Updated the RFC to simplify the wording of this part

infinisil commented:
I updated this PR to reflect the most recent discussions we've been having and addressing all reviews. Please review again @roberth, @SuperSandro2000 but also others in the @nixpkgs-architecture/team.

roberth commented on README.md:
```diff
@@ -11,78 +11,114 @@ related-issues: (will contain links to implementation PRs)
+    - On GitHub this is even more problematic, as the `all-packages.nix` file is [too big to be displayed by GitHub](https://github.com/NixOS/nixpkgs/blob/nixos-22.05/pkgs/top-level/all-packages.nix)
   - Then go to that file's definition, which takes quite some time for navigation (unless you have a plugin that can jump to it directly)
+    - It also slows down or even deadlocks editors due to the file size
+  - In some cases `nix edit` works, though that's not yet stable (it relies on Flakes being enabled) and comes with some problems ([doesn't yet open a writable file](https://github.com/NixOS/nix/issues/3347), doesn't work with packages that don't set `meta.position` correctly).
```
```suggestion
  - `nix edit -f . package-attr` works, though that's not yet stable (it relies on the `nix-command` feature being enabled) and doesn't work with packages that don't set `meta.position` correctly).
```

`nix edit` existed before flakes iirc and it works.

```
~/nixpkgs$ nix edit -f . hello --extra-experimental-features nix-command
```

roberth commented on README.md:
```diff
@@ -11,78 +11,114 @@ related-issues: (will contain links to implementation PRs)
+1. <a id="criteria-1"/> Is defined in `pkgs/top-level/all-packages.nix`
+  (necessary so that the overlay containing the automatically discovered packages can be ordered directly before the `all-packages.nix` overlay without changing any behavior)
+2. <a id="criteria-2"/> Is defined to be equal to `pkgs.callPackage <path> { }`
+3. <a id="criteria-3"/> All transitively referenced paths from the default Nix file of `<path>` are under the same directory as the default Nix file and can be moved around together without breaking any references in other Nix files (except the one reference in `pkgs/top-level/all-packages.nix`).
```
This is an implementation detail, not an intentional restriction.

Please separate the end goal from what the current tool can do.

infinisil replied:
I guess you're thinking of the tool doing automatic fixing of references in Nix files when moving paths around? At least in all the cases I've seen, that's a violation of API boundaries: Packages shouldn't rely on paths outside of their own "unit" directory. And similarly the other way around. So as far as I can see, it's reasonable that the tool only supports this, and not just for the sake of simplifying the implementation.

roberth replied:
I think the reader would like to know in a bit more detail what Nixpkgs will look like before diving into the technical details of the process of getting there.

> I guess you're thinking of the tool doing automatic fixing of references in Nix files when moving paths around?

That would be nice, but isn't really the point.

> >  implementation detail, not an intentional restriction

Suppose I want to move `cassandra` into `pkgs/unit`. Its files are referenced by `pkgs.cassandra` and `pkgs.cassandra_3_11`. Am I not allowed to move it, because it violates (2) and (3)?

infinisil replied:
I rewrote the details section a bunch to not mention implementation details anymore.

roberth commented on README.md:
```diff
@@ -11,78 +11,114 @@ related-issues: (will contain links to implementation PRs)
+
+These attributes will newly be added to `pkgs` by automatically calling `pkgs.callPackage pkgs/unit/<4-prefix name>/<name>/pkg-fun.nix { }` on all entries in `pkgs/unit`. In order to ensure efficiency of this operation, `builtins.readDir` should be optimized as described [here](https://github.com/NixOS/nix/issues/7314).
 
 ## Transitioning
```
```suggestion
## Transition
```

roberth commented on README.md:
```diff
@@ -11,78 +11,114 @@ related-issues: (will contain links to implementation PRs)
-- Rip-off-the-bandaid approach: Do a big transition, causing a lot of merge conflicts with existing PR's creating and updating packages. Inform PR authors on how to progress
-- Incremental CI approach: Introduce a CI action that complains when new packages (and maybe updated packages as well) use the old package paths
-- Transparent and sneaky approach: Transition only packages that can be done without causing merge conflicts for all existing PRs. Repeat every once in a while until all packages are done
+This RFC comes with [a reference tool](https://github.com/nixpkgs-architecture/simple-package-paths/pull/22) to make the above transition in an automated way.
```
```suggestion
This RFC comes with [a tool](https://github.com/nixpkgs-architecture/simple-package-paths/pull/22) to make the above transition in an automated way.
```

roberth commented on README.md:
```diff
@@ -11,78 +11,114 @@ related-issues: (will contain links to implementation PRs)
 [summary]: #summary
 
-Make most attribute definitions in `pkgs/top-level/all-packages.nix` be auto-generated from a single flat directory instead, where the subdirectory corresponds to the attribute name. The ad-hoc category-based structure of packages will be gotten rid of.
+Make trivial top-level attribute definitions in `pkgs/top-level/all-packages.nix` be auto-generated from a predictable attribute-based file hierarchy.
```
```suggestion
Auto-generate trivial top-level attribute definitions in `pkgs/top-level/all-packages.nix`  from a sharded directory that matches the attribute name.
```

 - Simplify sentence structure.
 - Predictable hierarchy suggests actual hierarchy. I don't like adding sharded here, but just _directory_ would be equally wrong, compared to _hierarchy_.

SuperSandro2000 reviewed (commented):
I don't have anything to add right now

roberth commented on README.md:
```diff
@@ -11,78 +11,114 @@ related-issues: (will contain links to implementation PRs)
+
+This RFC comes with [a tool](https://github.com/nixpkgs-architecture/simple-package-paths/pull/22) to make the above transition in an automated way.
+If this RFC is accepted, the result of that tool will be PR'd to nixpkgs.
+The tool itself will also be added to nixpkgs so that it can easily be ran again in the future.
```
```suggestion
The tool itself will also be added to nixpkgs so that it can easily be run again in the future.
```

"ran" is only simple past.

roberth commented on README.md:
```diff
@@ -11,78 +11,114 @@ related-issues: (will contain links to implementation PRs)
+## Transition
+
+This RFC comes with [a tool](https://github.com/nixpkgs-architecture/simple-package-paths/pull/22) to make the above transition in an automated way.
+If this RFC is accepted, the result of that tool will be PR'd to nixpkgs.
```
```suggestion
If this RFC is accepted, the result of that tool will be used to create a pull request to nixpkgs.
```

roberth commented on README.md:
```diff
@@ -11,78 +11,114 @@ related-issues: (will contain links to implementation PRs)
+This RFC comes with [a tool](https://github.com/nixpkgs-architecture/simple-package-paths/pull/22) to make the above transition in an automated way.
+If this RFC is accepted, the result of that tool will be PR'd to nixpkgs.
+The tool itself will also be added to nixpkgs so that it can easily be ran again in the future.
+For at least one release cycle, the legacy way of declaring packages should still be accepted, but the tool can be ran again at any point, thereby moving those new packages from the legacy paths to the new `pkgs/unit` paths.
```
```suggestion
For at least one release cycle, the legacy way of declaring packages should still be accepted, but the tool can be run again at any point, thereby moving those new packages from the legacy paths to the new `pkgs/unit` paths.
```

roberth commented on README.md:
```diff
@@ -11,78 +11,114 @@ related-issues: (will contain links to implementation PRs)
 
-- `nix edit` is unaffected, since it uses a packages `meta.position` to get the file to edit
+- `nix edit` is unaffected, since it uses a packages `meta.position` to get the file to edit.
+  Though with this RFC `nix edit` could be updated to not have to rely on that anymore for the packages in the new hierarchy in nixpkgs.
```
```suggestion
```

I don't think it can or should.

infinisil replied:
If nixpkgs sufficiently exposes it as a stable API, I think it could. The benefit would be that `nix edit` could also work on packages where `meta.position` can't be evaluated, e.g. due to syntax errors. Additionally, `nix edit` might open the wrong file due to a builder not properly populating `meta.position`, this problem would also be avoided.

infinisil replied:
We discussed this [in the last NAT meeting](https://discourse.nixos.org/t/2022-12-19-nixpkgs-architecture-team-meeting-22/24125#httpsgithubcomnixpkgs-architecturesimple-package-pathspull20discussion_r1051188387-4), I removed the sentence now

roberth commented on README.md:
```diff
@@ -93,52 +129,51 @@ This RFC makes no requirement as to how the transition should happen, but here a
+  - Good because it's simpler, both for the user and for the code
+  - Good because it speeds up Nix evaluation since there's only a single directory to call `builtins.readDir` on instead of many
+    - With an optimized `readDir` this isn't much of a problem
+  - Bad because it causes GitHub to limit the rendering of that directory to 1'000 entries (and we have about 10'000 that benefit from this transition for a start)
```
```suggestion
  - Bad because the GitHub web interface only renders the first 1'000 entries (and we have about 10'000 that benefit from this transition, even given the restrictions)
```

roberth commented on README.md:
```diff
@@ -93,52 +129,51 @@ This RFC makes no requirement as to how the transition should happen, but here a
+  - Bad because directory listing slows down with many files
+
+- Don't use `pkg-fun.nix` but another file name:
+  - `package.nix`/`pkg.nix`: Makes the migration to a non-function form of overridable packages harder in the future.
```
```suggestion
  - `package.nix`/`pkg.nix`: Bad, because it makes the migration to a non-function form of overridable packages harder in the future.
```

roberth replied:
We should rely on an implied meaning of the bullets below either.

roberth commented on README.md:
```diff
@@ -93,52 +129,51 @@ This RFC makes no requirement as to how the transition should happen, but here a
+  This isn't done because it [turns out](https://github.com/nixpkgs-architecture/simple-package-paths/issues/14) that this criteria indicates the file structure being used as an API interface.
+  By manually refactoring the Nix code to not rely on this anymore, you can increase code quality/reusability/clarity and then do the transition described in the RFC.
+- Use a different sharding scheme than `<4-prefix name>`.
+  Discussions regarding this can be seen [here](https://github.com/nixpkgs-architecture/simple-package-paths/issues/1), [NAT meeting #18](https://github.com/nixpkgs-architecture/meetings/blob/6282b0c6bbc47b6f1becd155586c79728eddefc9/2022-11-21.md) and [here](https://github.com/nixpkgs-architecture/simple-package-paths/pull/20#discussion_r1029004083)
```
Should be summarized here.

infinisil replied:
I redid the alternatives section, including writing a summary of this one here

roberth commented on README.md:
```diff
@@ -93,52 +129,51 @@ This RFC makes no requirement as to how the transition should happen, but here a
 
+All of these questions are in scope to be addressed in future discussions in the [Nixpkgs Architecture Team](https://nixos.org/community/teams/nixpkgs-architecture.html):
+
 - This RFC only addresses the top-level attribute namespace, aka packages in `pkgs.<name>`, it doesn't do anything about package sets like `pkgs.python3Packages.<name>`, `pkgs.haskell.packages.ghc942.<name>`, which could also benefit from a similar auto-calling
```
```suggestion
- This RFC only addresses the top-level attribute namespace, aka packages in `pkgs.<name>`, it doesn't do anything about package sets like `pkgs.python3Packages.<name>`, `pkgs.haskell.packages.ghc942.<name>`, which may or may not benefit from a similar auto-calling
```

nixos-discourse commented:
This pull request has been mentioned on **NixOS Discourse**. There might be relevant details there:

https://discourse.nixos.org/t/2022-12-19-nixpkgs-architecture-team-meeting-22/24125/1

infinisil commented on README.md:
```diff
@@ -11,78 +11,83 @@ related-issues: (will contain links to implementation PRs)
 
-## Backwards compatibility symlinks
-[symlinks]: #backwards-compatibility-symlinks
+These requirements will be checked using CI:
+1. <a id="req-1"/> The `pkgs/unit` directory must only contain unit directories, and only in subdirectories of the form `${substring 0 4 name}/${name}`.
+2. <a id="req-2"/> Files outside a unit directory must not reference files inside that unit directory, and the other way around.
+3. <a id="req-3"/> Each unit directory must have a `pkg-fun.nix` file such that `pkgs.callPackage ./pkg-fun.nix {}` evaluates to a derivation.
+4. <a id="req-4"/> Packages defined by unit directories must not be defined or overridden anywhere else, such as in `pkgs/top-level/all-packages.nix`.
```
I also codified these requirements that should be checked by CI

1. So that people don't pollute the `pkgs/unit` directory with wrong shards
2. Makes the unit self-contained which is a good thing to have in general, and we get this check essentially for free with the tooling for the transition. If this becomes a point of contention we could also remove this requirement though.
3. So we can always `callPackage`
4. So that we don't have one person adding a file in `all-packages.nix` and another in `pkgs/unit`, and one of them being ignored without any indication.

infinisil commented on README.md:
```diff
@@ -11,78 +11,83 @@ related-issues: (will contain links to implementation PRs)
+1. The `pkgs/unit` directory must only contain unit directories, and only in subdirectories of the form `${substring 0 4 name}/${name}`.
+2. <a id="req-ref"/> Files outside a unit directory must not reference files inside that unit directory, and the other way around.
+3. Each unit directory must have a `pkg-fun.nix` file such that `pkgs.callPackage ./pkg-fun.nix {}` evaluates to a derivation.
+4. Packages defined by unit directories must not be defined or overridden anywhere else, such as in `pkgs/top-level/all-packages.nix`.
```
```suggestion
4. The definition of a package in the unit directory is the one `pkgs.<name>` points to.
5. To avoid problems with merges, if a package attribute is defined by a unit directory, it must not be defined in `pkgs/top-level/all-packages.nix` or `pkgs/top-level/aliases.nix`.
```

infinisil commented on README.md:
```diff
@@ -11,78 +11,83 @@ related-issues: (will contain links to implementation PRs)
+These requirements will be checked using CI:
+1. The `pkgs/unit` directory must only contain unit directories, and only in subdirectories of the form `${substring 0 4 name}/${name}`.
+2. <a id="req-ref"/> Files outside a unit directory must not reference files inside that unit directory, and the other way around.
+3. Each unit directory must have a `pkg-fun.nix` file such that `pkgs.callPackage ./pkg-fun.nix {}` evaluates to a derivation.
```
```suggestion
```

Ericson2314 commented on README.md:
```diff
@@ -93,53 +98,68 @@ This RFC makes no requirement as to how the transition should happen, but here a
+
+## Not having the [reference requirement](#user-content-req-ref)
+
+The reference requirement could be removed, which would allow unit directories to reference files outside themselves, and the other way around. This is not great because it encourages the use of file paths as an API, rather than explicitly exposing functionality from Nix expressions.
```
```suggestion
The reference requirement could be removed, which would allow unit directories to reference files outside themselves, and the other way around. This is not great because it encourages the use of file paths as an API, rather than explicitly exposing functionality from Nix expressions.

## Relax design to try to attack issues like "package variants" up front

An issue with restrictions like the above one is that they don't work well for when we package a number of variants of package, e.g. different versions of the same package that share some infra. We do presume we would have to have *some* notion of "private details shared between multiple units" or "multiple entry points to unit" to handle these cases.

We've chosen to explicitly ignore these tough cases, and emphasize uniform structure of units over being able to migrate over as many packages as possible from the get go. The rationale for this decision is basically:

1. It is (a bit) easier to relax requirements later than tighten them later.
2. We plan on incrementally migrating Nixpkgs to this new system anyways, for caution's sake, so starting with fewer units is not only fine but *good*.
3. Explicitly marking use-cases out of scope allows us to have a more focused and thorough investigation of the use-cases that remain, to build a solid foundation.
```

nixos-discourse commented:
This pull request has been mentioned on **NixOS Discourse**. There might be relevant details there:

https://discourse.nixos.org/t/2022-01-03-nixpkgs-architecture-team-meeting-23/24404/1
